src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+            const counts_size = @divExact(value_count_max * layout.clock_bits, 8);
+            assert(counts_size >= layout.cache_line_size);
+            assert(counts_size % layout.cache_line_size == 0);
+
+            const clocks_size = @divExact(sets * clock_hand_bits, 8);
+            assert(clocks_size >= layout.cache_line_size);
+            assert(clocks_size % layout.cache_line_size == 0);
+
+            const tags = try allocator.alloc(Tag, value_count_max);
+            errdefer allocator.free(tags);
+
+            const values = try allocator.allocAdvanced(
+                Value,
+                value_alignment,
+                value_count_max,
+                .exact,
+            );
+            errdefer allocator.free(values);
+
+            const counts = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(counts);
+
+            const clocks = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(clocks);
+
+            _ = Clocks;
+            _ = Counts;
+
+            var self = Self{
+                .sets = sets,
+                .tags = tags,
+                .values = values,
+                .counts = counts,
+                .clocks = clocks,
+            };
+
+            self.reset();
+
+            return self;
+        }
+
+        pub fn deinit(self: *Self, allocator: mem.Allocator) void {
+            assert(self.sets > 0);
+            self.sets = 0;
+
+            allocator.free(self.tags);
+            allocator.free(self.values);
+            allocator.free(self.counts);
+            allocator.free(self.clocks);
+        }
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts, 0);
+            mem.set(u64, self.clocks, 0);
+        }
+
+        pub fn get(self: *Self, key: Key) ?*align(value_alignment) Value {
+            const set = self.associate(key);
+            return self.search(set, key);
+        }
+
+        pub const GetOrPutResult = struct {
+            value_ptr: *align(value_alignment) Value,
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
+            return self.get_or_put_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Get a pointer to an existing value with the given key if the key is in the cache.
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
```

Comment:
It would be great to have a unit test for this function, checking how the clock hand is advanced across calls to the function, and how locked slots are skipped.